### PR TITLE
Enable Go race detector in CI

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
- [ ] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [ ] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [ ] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
Fixes https://github.com/Azure/azure-sdk-for-go/issues/16112

File: sdk/messaging/azeventhubs/internal/namespace_test.go [hunk 1]
```diff
@@ -118,7 +118,7 @@ func TestNamespaceNegotiateClaimRenewal(t *testing.T) {
 	}
 
 	var errorsLogged []error
-	var nextRefreshDurationChecks atomic.Int32
+	var nextRefreshDurationChecks int32
```
Comment (richardpark-msft) [inline sdk/messaging/azeventhubs/internal/namespace_test.go:121]: Oh, I know this "downgrade" feeling. They added some nice stuff in atomics in 1.19.

Comment (jhendrixMSFT) [reply sdk/messaging/azeventhubs/internal/namespace_test.go:121]: Yeah I was a bit bummed about having to make this switch.

**Comment by jhendrixMSFT** [comment]
@benbp enabling race detection has required certain tests to be excluded due to the tests failing with out-of-memory errors (this isn't surprising due to how the race detector works and the tests allocating large amounts of memory), and since we want race detection on by default, it means the impacted tests won't run in CI which isn't ideal.

What I'm thinking is that we add some extra flag in `ci.yml` files that indicates "please also run tests in the enumerated directories without race detection enabled".  Does that seem reasonable/doable?  Open to other ideas.

**Comment by jhendrixMSFT** [comment]
Here's an example.

https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2661578&view=logs&j=a4d06d16-879e-5ba3-ceb1-b266dbe8989a&t=35eb6bb5-29cf-53ca-cab6-34e1cf74a298

It's easiest to see this on the Windows runs, here's the relevant text (for Linux, the tests appear to just die without any info).
```
2023-03-28T19:15:13.8639015Z ==5880==ERROR: ThreadSanitizer failed to allocate 0x000274000000 (10536091648) bytes at 0x040477000000 (error code: 1455)
```

Error code 1455 means the [page file is too small to complete the operation](https://learn.microsoft.com/en-us/windows/win32/debug/system-error-codes--1300-1699-).

The exclusions are at the test level by way of a build tag, see file `client_norace_test.go` in this PR.  So all tests in this file will be excluded from running with the race detector.

I think the easiest/safest thing to do would be to run tests in the specified directories without the race detector.  So in this case, we'd run the tests in `blockblob` twice.  While it does mean there will be some duplication, I think going through and annotating every other test file (and remembering to do so in the future) just makes things too complicated.

**Comment by benbp** [comment]
@jhendrixMSFT Yeah the annotation is not ideal but I also don't think that doubling the test time is great either. Is this just the blob tests (e.g. they consume more memory because they upload/download large blobs and that affects the race detector) or is it more widespread?

**Comment by jhendrixMSFT** [comment]
At present, it's just the handful of tests I moved to `client_norace_test.go` as they allocate large chunks of memory for large blobs.

**Pull request closed without merging**